ethereum/eth/src/main/java/tech/pegasys/pantheon/ethereum/eth/sync/worldstate/WorldStateDownloader.java

```diff
@@ -146,31 +93,52 @@ public WorldStateDownloader(
 
       final Hash stateRoot = header.getStateRoot();
       if (worldStateStorage.isWorldStateAvailable(stateRoot)) {
+        LOG.info(
+            "World state already available for block {} ({}). State root {}",
+            header.getNumber(),
+            header.getHash(),
+            stateRoot);
         return CompletableFuture.completedFuture(null);
       }
+      LOG.info(
+          "Begin downloading world state from peers for block {} ({}). State root {}",
+          header.getNumber(),
+          header.getHash(),
+          stateRoot);
 
-      // Room for the requests we expect to do in parallel plus some buffer but not unlimited.
-      final int persistenceQueueCapacity = hashCountPerRequest * maxOutstandingRequests * 2;
       final WorldDownloadState newDownloadState =
-          new WorldDownloadState(
-              taskCollection,
-              new ArrayBlockingQueue<>(persistenceQueueCapacity),
-              maxOutstandingRequests,
-              maxNodeRequestsWithoutProgress);
+          new WorldDownloadState(taskCollection, maxNodeRequestsWithoutProgress);
       this.downloadState.set(newDownloadState);
 
       if (!newDownloadState.downloadWasResumed()) {
         // Only queue the root node if we're starting a new download from scratch
         newDownloadState.enqueueRequest(NodeDataRequest.createAccountDataRequest(stateRoot));
       }
 
-      ethContext
-          .getScheduler()
-          .scheduleSyncWorkerTask(() -> requestNodeData(header, newDownloadState));
-
-      final PersistNodeDataTask persistenceTask = new PersistNodeDataTask(header, newDownloadState);
-      newDownloadState.setPersistenceTask(persistenceTask);
-      ethContext.getScheduler().scheduleServiceTask(persistenceTask);
+      final WorldStateDownloadProcess downloadProcess =
+          WorldStateDownloadProcess.builder()
+              .hashCountPerRequest(hashCountPerRequest)
+              .maxOutstandingRequests(maxOutstandingRequests)
+              .loadLocalDataStep(new LoadLocalDataStep(worldStateStorage, metricsSystem))
+              .requestDataStep(new RequestDataStep(ethContext, metricsSystem))
+              .persistDataStep(new PersistDataStep(worldStateStorage))
+              .completeTaskStep(new CompleteTaskStep(worldStateStorage, metricsSystem))
+              .downloadState(newDownloadState)
+              .pivotBlockHeader(header)
+              .metricsSystem(metricsSystem)
+              .build();
+
+      newDownloadState.setWorldStateDownloadProcess(downloadProcess);
+
+      final CompletableFuture<Void> downloadProcessFuture =
+          downloadProcess.start(ethContext.getScheduler());
+      downloadProcessFuture.whenComplete(
+          (result, error) -> {
+            if (error != null
+                && !(ExceptionUtils.rootCause(error) instanceof CancellationException)) {
```

Comment:
If the pipeline was cancelled it's because the download state cancelled so we don't propagate the cancellation or we'll interfere with the clean up doing. Typically that results in the fast sync being considered as failed when shutting down the client so it deletes its state.

I've added a comment explaining that but may also play with a follow up PR to move starting the pipeline into `WorldDownloadState` so that it winds up managing the future itself and avoid having the logic split over two classes.